File: api/project.py
import time
import traceback
import asyncio
from bson import ObjectId
from fastapi import APIRouter, Depends, BackgroundTasks

from api.task import create_scan_task, delete_asset
from api.users import verify_token
from motor.motor_asyncio import AsyncIOMotorCursor

from core.config import Project_List
from core.db import get_mongo_db
from core.redis_handler import refresh_config, get_redis_pool
from loguru import logger
from core.util import *
from core.apscheduler_handler import scheduler

router = APIRouter()


@router.post("/project/data")
async def get_projects_data(request_data: dict, db=Depends(get_mongo_db), _: dict = Depends(verify_token),
                            background_tasks: BackgroundTasks = BackgroundTasks()):
    background_tasks.add_task(update_project_count)
    search_query = request_data.get("search", "")
    page_index = request_data.get("pageIndex", 1)
    page_size = request_data.get("pageSize", 10)

    query = {
        "$or": [
            {"name": {"$regex": search_query, "$options": "i"}},
            {"root_domain": {"$regex": search_query, "$options": "i"}}
        ]
    } if search_query else {}

    # 获取标签统计信息
    tag_result = await db.project.aggregate([
        {"$group": {"_id": "$tag", "count": {"$sum": 1}}},
        {"$sort": {"count": -1}}
    ]).to_list(None)

    tag_num = {tag["_id"]: tag["count"] for tag in tag_result}
    all_num = sum(tag_num.values())
    tag_num["All"] = all_num

    result_list = {}

    async def fetch_projects(tag, tag_query):
        cursor = db.project.find(tag_query, {
            "_id": 0,
            "id": {"$toString": "$_id"},
            "name": 1,
            "logo": 1,
            "AssetCount": 1,
            "tag": 1
        }).sort("AssetCount", -1).skip((page_index - 1) * page_size).limit(page_size)

        results = await cursor.to_list(length=None)
        for result in results:
            result["AssetCount"] = result.get("AssetCount", 0)
        return results

    fetch_tasks = []
    for tag in tag_num:
        if tag != "All":
            tag_query = {"$and": [query, {"tag": tag}]}
        else:
            tag_query = query

        fetch_tasks.append(fetch_projects(tag, tag_query))

    fetch_results = await asyncio.gather(*fetch_tasks)

    for tag, results in zip(tag_num, fetch_results):
        result_list[tag] = results

    return {
        "code": 200,
        "data": {
            "result": result_list,
            "tag": tag_num
        }
    }


@router.get("/project/all")
async def get_projects_all(db=Depends(get_mongo_db), _: dict = Depends(verify_token)):
    try:
        pipeline = [
            {
                "$group": {
                    "_id": "$tag",  # 根据 tag 字段分组
                    "children": {"$push": {"value": {"$toString": "$_id"}, "label": "$name"}}  # 将每个文档的 _id 和 name 放入 children 集合中
                }
            },
            {
                "$project": {
                    "_id": 0,
                    "label": "$_id",
                    "value": {"$literal": ""},
                    "children": 1
                }
            }
        ]
        result = await db['project'].aggregate(pipeline).to_list(None)
        return {
            "code": 200,
            "data": {
                'list': result
            }
        }
    except Exception as e:
        logger.error(str(e))
        logger.error(traceback.format_exc())
        return {"message": "error","code":500}


async def update_project_count():
    async for db in get_mongo_db():
        cursor = db.project.find({}, {"_id": 0, "id": {"$toString": "$_id"}})
        results = await cursor.to_list(length=None)

        async def update_count(id):
            query = {"project": {"$eq": id}}
            total_count = await db.asset.count_documents(query)
            if total_count != 0:
                update_document = {
                    "$set": {
                        "AssetCount": total_count
                    }
                }
                await db.project.update_one({"_id": ObjectId(id)}, update_document)

        fetch_tasks = [update_count(r['id']) for r in results]

        await asyncio.gather(*fetch_tasks)


@router.post("/project/content")
async def get_project_content(request_data: dict, db=Depends(get_mongo_db), _: dict = Depends(verify_token)):
    project_id = request_data.get("id")
    if not project_id:
        return {"message": "ID is missing in the request data", "code": 400}
    query = {"_id": ObjectId(project_id)}
    doc = await db.project.find_one(query)
    if not doc:
        return {"message": "Content not found for the provided ID", "code": 404}
    project_target_data = await db.ProjectTargetData.find_one({"id": project_id})
    result = {
        "name": doc.get("name", ""),
        "tag": doc.get("tag", ""),
        "target": project_target_data.get("target", ""),
        "node": doc.get("node", []),
        "logo": doc.get("logo", ""),
        "subdomainScan": doc.get("subdomainScan", False),
        "subdomainConfig": doc.get("subdomainConfig", []),
        "urlScan": doc.get("urlScan", False),
        "sensitiveInfoScan": doc.get("sensitiveInfoScan", False),
        "pageMonitoring": doc.get("pageMonitoring", ""),
        "crawlerScan": doc.get("crawlerScan", False),
        "vulScan": doc.get("vulScan", False),
        "vulList": doc.get("vulList", []),
        "portScan": doc.get("portScan"),
        "ports": doc.get("ports"),
        "waybackurl": doc.get("waybackurl"),
        "dirScan": doc.get("dirScan"),
        "scheduledTasks": doc.get("scheduledTasks"),
        "hour": doc.get("hour"),
        "allNode": doc.get("allNode", False),
        "duplicates": doc.get("duplicates")
    }
    return {"code": 200, "data": result}


@router.post("/project/add")
async def add_project_rule(request_data: dict, db=Depends(get_mongo_db), _: dict = Depends(verify_token),
                           background_tasks: BackgroundTasks = BackgroundTasks()):
    try:
        # Extract values from request data
        name = request_data.get("name")
        target = request_data.get("target")
        runNow = request_data.get("runNow")
        del request_data["runNow"]
        scheduledTasks = request_data.get("scheduledTasks", False)
        hour = request_data.get("hour", 1)
        t_list = []
        tmsg = ''
        root_domains = []
        for t in target.split('\n'):
            if t not in t_list:
                targetTmp = t.replace('http://', "").replace('https://', "").strip()
                if targetTmp != "":
                    root_domain = get_root_domain(targetTmp)
                    if root_domain not in root_domains:
                        root_domains.append(root_domain)
                    tmsg += targetTmp + '\n'
                # Create a new SensitiveRule document
        tmsg = tmsg.strip().strip("\n")
        request_data["root_domains"] = root_domains
        del request_data['target']
        if "All Poc" in request_data['vulList']:
            request_data['vulList'] = ["All Poc"]
        cursor = db.project.find({"name": {"$eq": name}}, {"_id": 1})
        results = await cursor.to_list(length=None)
        if len(results) != 0:
            return {"code": 400, "message": "name already exists"}
        # Insert the new document into the SensitiveRule collection
        result = await db.project.insert_one(request_data)
        # Check if the insertion was successful
        if result.inserted_id:
            await db.ProjectTargetData.insert_one({"id": str(result.inserted_id), "target": tmsg})
            if scheduledTasks:
                scheduler.add_job(scheduler_project, 'interval', hours=hour, args=[str(result.inserted_id)],
                                  id=str(result.inserted_id), jobstore='mongo')
                next_time = scheduler.get_job(str(result.inserted_id)).next_run_time
                formatted_time = next_time.strftime("%Y-%m-%d %H:%M:%S")
                db.ScheduledTasks.insert_one(
                    {"id": str(result.inserted_id), "name": name, 'hour': hour, 'type': 'Project', 'state': True,
                     'lastTime': get_now_time(), 'nextTime': formatted_time, 'runner_id': str(result.inserted_id)})
                if runNow:
                    await scheduler_project(str(result.inserted_id))
            background_tasks.add_task(update_project, root_domains, str(result.inserted_id))
            await refresh_config('all', 'project')
            Project_List[name] = str(result.inserted_id)
            return {"code": 200, "message": "Project added successfully"}
        else:
            return {"code": 400, "message": "Failed to add Project"}

    except Exception as e:
        logger.error(str(e))
        # Handle exceptions as needed
        return {"message": "error", "code": 500}


@router.post("/project/delete")
async def delete_project_rules(request_data: dict, db=Depends(get_mongo_db), _: dict = Depends(verify_token),
                               background_tasks: BackgroundTasks = BackgroundTasks()):
    try:
        pro_ids = request_data.get("ids", [])
        delA = request_data.get("delA", False)
        if delA:
            background_tasks.add_task(delete_asset, pro_ids, True)
        obj_ids = [ObjectId(poc_id) for poc_id in pro_ids]
        result = await db.project.delete_many({"_id": {"$in": obj_ids}})
        await db.ProjectTargetData.delete_many({"id": {"$in": pro_ids}})
        # Check if the deletion was successful
        if result.deleted_count > 0:
            for pro_id in pro_ids:
                job = scheduler.get_job(pro_id)
                if job:
                    scheduler.remove_job(pro_id)
                background_tasks.add_task(delete_asset_project_handler, pro_id)
                for project_id in Project_List:
                    if pro_id == Project_List[project_id]:
                        del Project_List[project_id]
                        break
            await db.ScheduledTasks.delete_many({"id": {"$in": pro_ids}})
            return {"code": 200, "message": "Project deleted successfully"}
        else:
            return {"code": 404, "message": "Project not found"}

    except Exception as e:
        logger.error(str(e))
        # Handle exceptions as needed
        return {"message": "error", "code": 500}


@router.post("/project/update")
async def update_project_data(request_data: dict, db=Depends(get_mongo_db), _: dict = Depends(verify_token),
                              background_tasks: BackgroundTasks = BackgroundTasks()):
    try:
        # Get the ID from the request data
        pro_id = request_data.get("id")
        hour = request_data.get("hour")
        runNow = request_data.get("runNow")
        del request_data["runNow"]
        # Check if ID is provided
        if not pro_id:
            return {"message": "ID is missing in the request data", "code": 400}
        query = {"id": pro_id}
        doc = await db.ScheduledTasks.find_one(query)

        newScheduledTasks = request_data.get("scheduledTasks")
        if doc:
            oldScheduledTasks = doc["state"]
            old_hour = doc["hour"]
            if oldScheduledTasks != newScheduledTasks:
                if newScheduledTasks:
                    scheduler.add_job(scheduler_project, 'interval', hours=hour, args=[pro_id],
                                      id=str(pro_id), jobstore='mongo')
                    await db.ScheduledTasks.update_one({"id": pro_id}, {"$set": {'state': True}})
                else:
                    job = scheduler.get_job(pro_id)
                    if job is not None:
                        scheduler.remove_job(pro_id)
                    await db.ScheduledTasks.update_one({"id": pro_id}, {"$set": {'state': False}})
            else:
                if newScheduledTasks:
                    if hour != old_hour:
                        job = scheduler.get_job(pro_id)
                        if job is not None:
                            scheduler.remove_job(pro_id)
                        scheduler.add_job(scheduler_project, 'interval', hours=hour, args=[pro_id],
                                          id=str(pro_id), jobstore='mongo')
        else:
            if newScheduledTasks:
                scheduler.add_job(scheduler_project, 'interval', hours=hour, args=[str(pro_id)],
                                  id=str(pro_id), jobstore='mongo')
                next_time = scheduler.get_job(str(pro_id)).next_run_time
                formatted_time = next_time.strftime("%Y-%m-%d %H:%M:%S")
                db.ScheduledTasks.insert_one(
                    {"id": str(pro_id), "name": request_data['name'], 'hour': hour, 'type': 'Project', 'state': True,
                     'lastTime': get_now_time(), 'nextTime': formatted_time})

        result = await db.project.find_one({"_id": ObjectId(pro_id)})
        project_target_data = await db.ProjectTargetData.find_one({"id": pro_id})
        old_targets = project_target_data['target']
        old_name = result['name']
        new_name = request_data['name']
        new_targets = request_data['target']
        if old_targets != new_targets.strip().strip('\n'):
            new_root_domain = []
            update_document = {
                "$set": {
                    "target": new_targets.strip().strip('\n')
                }
            }
            for n_t in new_targets.strip().strip('\n').split('\n'):
                t_root_domain = get_root_domain(n_t)
                if t_root_domain not in new_root_domain:
                    new_root_domain.append(t_root_domain)
            request_data["root_domains"] = new_root_domain
            await db.ProjectTargetData.update_one({"id": pro_id}, update_document)
            background_tasks.add_task(update_project, new_root_domain, pro_id, True)
        if old_name != new_name:
            del Project_List[old_name]
            Project_List[new_name] = pro_id
            await db.ScheduledTasks.update_one({"id": pro_id}, {"$set": {'name': new_name}})
        request_data.pop("id")
        del request_data['target']
        update_document = {
            "$set": request_data
        }
        result = await db.project.update_one({"_id": ObjectId(pro_id)}, update_document)
        # Check if the update was successful
        if result:
            if runNow:
                await scheduler_project(str(pro_id))
            return {"message": "Task updated successfully", "code": 200}
        else:
            return {"message": "Failed to update data", "code": 404}

    except Exception as e:
        logger.error(str(e))
        logger.error(traceback.format_exc())
        # Handle exceptions as needed
        return {"message": "error", "code": 500}

async def update_project(root_domain, project_id, change=False):
    asset_collection_list = {
                        'asset': ["url", "host", "ip"],
                        'subdomain': ["host", "ip"],
                        'DirScanResult': ["url"],
                        'vulnerability': ["url"],
                        'SubdoaminTakerResult': ["input"],
                        'PageMonitoring': ["url"],
                        'SensitiveResult': ["url"],
                        'UrlScan': ["input"],
                        'crawler': ["url"]}
    async for db in get_mongo_db():
        for a in asset_collection_list:
            if change:
                await asset_update_project(root_domain, asset_collection_list[a], a, db, project_id)
            else:
                await asset_add_project(root_domain, asset_collection_list[a], a, db, project_id)


async def asset_add_project(root_domain, db_key, doc_name, db, project_id):
    regex_patterns = [f".*{domain}.*" for domain in root_domain]
    pattern = "|".join(regex_patterns)
    # 构建查询条件
    query = {
        "$and": [
            {
                "$or": [
                    {key: {"$regex": pattern, "$options": "i"}} for key in db_key
                ]
            },
            {"project": {"$exists": True, "$eq": ""}}
        ]
    }
    update_query = {
        "$set": {
            "project": project_id
        }
    }
    result = await db[doc_name].update_many(query, update_query)
    # 打印更新的文档数量
    logger.info(f"Updated {doc_name} {result.modified_count} documents")


async def asset_update_project(root_domain, db_key, doc_name, db, project_id):
    regex_patterns = [f".*{domain}.*" for domain in root_domain]
    pattern = "|".join(regex_patterns)
    # 构建查询条件
    query = {
        "$and": [
            {"project": project_id},
            {
                "$nor": [
                    {key: {"$regex": pattern, "$options": "i"}} for key in db_key
                ]
            }
        ]
    }
    update_query = {
        "$set": {
            "project": ""
        }
    }
    result = await db[doc_name].update_many(query, update_query)
    # 打印更新的文档数量
    logger.info(f"Updated {doc_name} {result.modified_count} documents to null ")
    await asset_add_project(root_domain, db_key, doc_name, db, project_id)


async def delete_asset_project(db, collection, project_id):
    try:
        query = {"project": project_id}

        cursor = db[collection].find(query)

        async for document in cursor:
            await db[collection].update_one({"_id": document["_id"]}, {"$set": {"project": ""}})

    except Exception as e:
        logger.error(f"delete_asset_project error:{e}")


async def delete_asset_project_handler(project_id):
    async for db in get_mongo_db():
        asset_collection_list = ['asset', 'subdomain', 'DirScanResult', 'vulnerability', 'SubdoaminTakerResult',
                                 'PageMonitoring', 'SensitiveResult', 'UrlScan', 'crawler']
        for c in asset_collection_list:
            await delete_asset_project(db, c, project_id)


async def scheduler_project(id):
    logger.info(f"Scheduler project {id}")
    async for db in get_mongo_db():
        async for redis in get_redis_pool():
            job = scheduler.get_job(id)
            task_id = generate_random_string(15)
            if job:
                next_time = job.next_run_time
                formatted_time = next_time.strftime("%Y-%m-%d %H:%M:%S")
                doc = await db.ScheduledTasks.find_one({"id": id})
                run_id_last = doc.get("runner_id", "")
                if run_id_last != "":
                    progresskeys = await redis.keys(f"TaskInfo:progress:{run_id_last}:*")
                    for pgk in progresskeys:
                        await redis.delete(pgk)
                update_document = {
                    "$set": {
                        "lastTime": get_now_time(),
                        "nextTime": formatted_time,
                        "runner_id": task_id
                    }
                }
                await db.ScheduledTasks.update_one({"id": id}, update_document)
            query = {"_id": ObjectId(id)}
            doc = await db.project.find_one(query)
            targetList = []
            target_data = await db.ProjectTargetData.find_one({"id": id})
            for t in target_data.get('target', '').split("\n"):
                t.replace("http://", "").replace("https://", "")
                t = t.strip("\n").strip("\r").strip()
                if t != "" and t not in targetList:
                    targetList.append(t)
            await create_scan_task(doc, task_id, targetList, redis)

File: internal/repositories/project/project.go
package project

import (
	"context"
	"fmt"
	"github.com/Autumn-27/ScopeSentry/internal/constants"
	"strings"
	"unicode"

	"github.com/Autumn-27/ScopeSentry/internal/database/mongodb"
	"github.com/Autumn-27/ScopeSentry/internal/logger"
	"github.com/Autumn-27/ScopeSentry/internal/models"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
	"go.uber.org/zap"
)

type Repository interface {
	GetProjectsByTag(ctx context.Context, projection bson.M) ([]models.Project, error)
	GetTarget(ctx context.Context, id string) (string, error)
	GetTargets(ctx context.Context, ids []string) (string, error)

	AggregateTagCounts(ctx context.Context, search string) (map[string]int, error)
	ListProjectsByTag(ctx context.Context, search, tag string, pageIndex, pageSize int) ([]models.ProjectBrief, error)
	FindByID(ctx context.Context, id string) (*models.Project, error)
	ExistsByName(ctx context.Context, name string) (bool, error)
	InsertProject(ctx context.Context, p *models.Project) (string, error)
	UpsertProjectTarget(ctx context.Context, id string, target string) error
	CreateOrUpdateProjectSchedule(ctx context.Context, id string, hour int, state bool, name string) error
	RemoveProjectSchedule(ctx context.Context, id string) error
	DeleteProjects(ctx context.Context, ids []string) error
	DeleteProjectTargets(ctx context.Context, ids []string) error
	DeleteProjectAsset(ctx context.Context, ids []string) error
	UpdateProject(ctx context.Context, p *models.UpdateProject) error
	UpdateAssetsProject(ctx context.Context, rootDomains []string, projectID string, change bool) error
}

type repository struct {
	collection       *mongo.Collection
	targetCollection *mongo.Collection
	scheduled        *mongo.Collection
}

func NewRepository() Repository {
	return &repository{
		collection:       mongodb.DB.Collection("project"),
		targetCollection: mongodb.DB.Collection("ProjectTargetData"),
		scheduled:        mongodb.DB.Collection("ScheduledTasks"),
	}
}

func (r *repository) GetProjectsByTag(ctx context.Context, projection bson.M) ([]models.Project, error) {
	cursor, err := r.collection.Find(ctx, bson.M{}, options.Find().SetProjection(projection))
	if err != nil {
		return nil, fmt.Errorf("failed to find projects: %w", err)
	}
	defer cursor.Close(ctx)

	var projects []models.Project
	if err := cursor.All(ctx, &projects); err != nil {
		return nil, fmt.Errorf("failed to decode projects: %w", err)
	}

	return projects, nil
}

func (r *repository) GetTarget(ctx context.Context, id string) (string, error) {
	var result struct {
		Target string `bson:"target"`
	}
	err := r.targetCollection.FindOne(
		ctx,
		bson.M{"id": id},
		options.FindOne().SetProjection(bson.M{"target": 1, "_id": 0}),
	).Decode(&result)

	if err != nil {
		return "", fmt.Errorf("failed to get target: %w", err)
	}
	return result.Target, nil
}

func (r *repository) GetTargets(ctx context.Context, ids []string) (string, error) {
	cursor, err := r.targetCollection.Find(
		ctx,
		bson.M{"id": bson.M{"$in": ids}},
		options.Find().SetProjection(bson.M{"target": 1, "_id": 0}),
	)
	if err != nil {
		return "", fmt.Errorf("failed to find targets: %w", err)
	}
	defer cursor.Close(ctx)

	var results []struct {
		Target string `bson:"target"`
	}
	if err := cursor.All(ctx, &results); err != nil {
		return "", fmt.Errorf("failed to decode targets: %w", err)
	}

	// 将所有target拼接成一个字符串
	var builder strings.Builder
	for i, result := range results {
		builder.WriteString(strings.TrimRightFunc(result.Target, unicode.IsSpace))
		if i != len(results)-1 {
			builder.WriteString("\n")
		}
	}
	return builder.String(), nil
}

func (r *repository) AggregateTagCounts(ctx context.Context, search string) (map[string]int, error) {
	var pipeline mongo.Pipeline
	if search != "" {
		pipeline = append(pipeline, bson.D{{Key: "$match", Value: bson.M{"$or": []bson.M{
			{"name": bson.M{"$regex": search, "$options": "i"}},
			{"root_domain": bson.M{"$regex": search, "$options": "i"}},
		}}}})
	}
	pipeline = append(pipeline,
		bson.D{{Key: "$group", Value: bson.M{"_id": "$tag", "count": bson.M{"$sum": 1}}}},
		bson.D{{Key: "$sort", Value: bson.M{"count": -1}}},
	)
	cursor, err := r.collection.Aggregate(ctx, pipeline)
	if err != nil {
		return nil, err
	}
	defer cursor.Close(ctx)
	res := make(map[string]int)
	for cursor.Next(ctx) {
		var row struct {
			ID    string `bson:"_id"`
			Count int    `bson:"count"`
		}
		if err := cursor.Decode(&row); err != nil {
			return nil, err
		}
		res[row.ID] = row.Count
	}
	return res, nil
}

func (r *repository) ListProjectsByTag(ctx context.Context, search, tag string, pageIndex, pageSize int) ([]models.ProjectBrief, error) {
	filter := bson.M{}
	if search != "" {
		filter["$or"] = []bson.M{
			{"name": bson.M{"$regex": search, "$options": "i"}},
			{"root_domain": bson.M{"$regex": search, "$options": "i"}},
		}
	}
	if tag != "All" {
		filter = bson.M{"$and": []bson.M{filter, {"tag": tag}}}
	}
	opts := options.Find().
		SetProjection(bson.M{"_id": 1, "name": 1, "logo": 1, "AssetCount": 1, "tag": 1}).
		SetSort(bson.D{{Key: "AssetCount", Value: -1}}).
		SetSkip(int64((pageIndex - 1) * pageSize)).
		SetLimit(int64(pageSize))
	cursor, err := r.collection.Find(ctx, filter, opts)
	if err != nil {
		return nil, err
	}
	defer cursor.Close(ctx)
	var docs []struct {
		ID         primitive.ObjectID `bson:"_id"`
		Name       string             `bson:"name"`
		Logo       string             `bson:"logo"`
		AssetCount int                `bson:"AssetCount"`
		Tag        string             `bson:"tag"`
	}
	if err := cursor.All(ctx, &docs); err != nil {
		return nil, err
	}
	items := make([]models.ProjectBrief, 0, len(docs))
	for _, d := range docs {
		items = append(items, models.ProjectBrief{ID: d.ID.Hex(), Name: d.Name, Logo: d.Logo, AssetCount: d.AssetCount, Tag: d.Tag})
	}
	return items, nil
}

func (r *repository) FindByID(ctx context.Context, id string) (*models.Project, error) {
	obj, err := primitive.ObjectIDFromHex(id)
	if err != nil {
		return nil, nil
	}
	var p models.Project
	if err := r.collection.FindOne(ctx, bson.M{"_id": obj}).Decode(&p); err != nil {
		if err == mongo.ErrNoDocuments {
			return nil, nil
		}
		return nil, err
	}
	return &p, nil
}

func (r *repository) ExistsByName(ctx context.Context, name string) (bool, error) {
	cnt, err := r.collection.CountDocuments(ctx, bson.M{"name": name})
	if err != nil {
		return false, err
	}
	return cnt > 0, nil
}

func (r *repository) InsertProject(ctx context.Context, p *models.Project) (string, error) {
	res, err := r.collection.InsertOne(ctx, p)
	if err != nil {
		return "", err
	}
	id := res.InsertedID.(primitive.ObjectID).Hex()
	return id, nil
}

func (r *repository) UpsertProjectTarget(ctx context.Context, id string, target string) error {
	_, err := r.targetCollection.UpdateOne(ctx, bson.M{"id": id}, bson.M{"$set": bson.M{"target": target}}, options.Update().SetUpsert(true))
	return err
}

func (r *repository) CreateOrUpdateProjectSchedule(ctx context.Context, id string, hour int, state bool, name string) error {
	data := bson.M{"id": id, "hour": hour, "state": state, "name": name, "type": "project"}
	_, err := r.scheduled.UpdateOne(ctx, bson.M{"id": id}, bson.M{"$set": data}, options.Update().SetUpsert(true))
	return err
}

func (r *repository) RemoveProjectSchedule(ctx context.Context, id string) error {
	_, err := r.scheduled.DeleteMany(ctx, bson.M{"id": id})
	return err
}

func (r *repository) DeleteProjects(ctx context.Context, ids []string) error {
	objIDs := make([]primitive.ObjectID, 0, len(ids))
	for _, id := range ids {
		if o, err := primitive.ObjectIDFromHex(id); err == nil {
			objIDs = append(objIDs, o)
		}
	}
	_, err := r.collection.DeleteMany(ctx, bson.M{"_id": bson.M{"$in": objIDs}})
	return err
}

func (r *repository) DeleteProjectTargets(ctx context.Context, ids []string) error {
	_, err := r.targetCollection.DeleteMany(ctx, bson.M{"id": bson.M{"$in": ids}})
	return err
}

func (r *repository) UpdateProject(ctx context.Context, p *models.UpdateProject) error {
	objID, err := primitive.ObjectIDFromHex(p.ID)
	if err != nil {
		return fmt.Errorf("invalid project ID: %v", err)
	}

	filter := bson.M{"_id": objID}
	_, err = r.collection.UpdateOne(ctx, filter, bson.M{"$set": p})
	return err
}

func (r *repository) DeleteProjectAsset(ctx context.Context, ids []string) error {
	for _, collectionName := range constants.AssetDBNames {
		collection := mongodb.DB.Collection(collectionName)
		_, err := collection.DeleteMany(ctx, bson.M{"project": bson.M{"$in": ids}})
		if err != nil {
			logger.Error(err.Error())
			continue
		}
	}
	return nil
}

// UpdateAssetsProject 更新资产项目关联
func (r *repository) UpdateAssetsProject(ctx context.Context, rootDomains []string, projectID string, change bool) error {
	// 资产集合列表

	for _, collectionName := range constants.AssetDBNames {
		if change {
			err := r.assetUpdateProject(ctx, rootDomains, collectionName, projectID)
			if err != nil {
				return fmt.Errorf("failed to update assets for collection %s: %w", collectionName, err)
			}
		} else {
			err := r.assetAddProject(ctx, rootDomains, collectionName, projectID)
			if err != nil {
				return fmt.Errorf("failed to add assets for collection %s: %w", collectionName, err)
			}
		}
	}

	return nil
}

// assetAddProject 添加资产到项目
func (r *repository) assetAddProject(ctx context.Context, rootDomains []string, collectionName, projectID string) error {
	collection := mongodb.DB.Collection(collectionName)

	var query bson.M

	switch collectionName {
	case "RootDomain":
		query = r.buildRootDomainQuery(rootDomains, false)
	case "app":
		query = r.buildAppQuery(rootDomains, false)
	case "mp":
		query = r.buildMpQuery(rootDomains, false)
	default:
		query = bson.M{"rootDomain": bson.M{"$in": rootDomains}}
	}

	updateQuery := bson.M{"$set": bson.M{"project": projectID}}

	result, err := collection.UpdateMany(ctx, query, updateQuery)
	if err != nil {
		return err
	}

	logger.Info("Updated assets for collection",
		zap.String("collection", collectionName),
		zap.Int64("modified_count", result.ModifiedCount))
	return nil
}

// assetUpdateProject 更新资产项目关联
func (r *repository) assetUpdateProject(ctx context.Context, rootDomains []string, collectionName, projectID string) error {
	collection := mongodb.DB.Collection(collectionName)

	var query bson.M

	switch collectionName {
	case "RootDomain":
		query = r.buildRootDomainQuery(rootDomains, true)
	case "app":
		query = r.buildAppQuery(rootDomains, true)
	case "mp":
		query = r.buildMpQuery(rootDomains, true)
	default:
		query = bson.M{
			"$and": []bson.M{
				{"project": projectID},
				{"rootDomain": bson.M{"$nin": rootDomains}},
			},
		}
	}

	// 先清空不符合条件的项目关联
	updateQuery := bson.M{"$set": bson.M{"project": ""}}
	result, err := collection.UpdateMany(ctx, query, updateQuery)
	if err != nil {
		return err
	}

	logger.Info("Cleared assets project association",
		zap.String("collection", collectionName),
		zap.Int64("modified_count", result.ModifiedCount))

	// 然后添加符合条件的项目关联
	return r.assetAddProject(ctx, rootDomains, collectionName, projectID)
}

// buildRootDomainQuery 构建RootDomain查询条件
func (r *repository) buildRootDomainQuery(rootDomains []string, isUpdate bool) bson.M {
	regexPatterns := make([]string, len(rootDomains))
	for i, domain := range rootDomains {
		regexPatterns[i] = "^" + strings.ReplaceAll(domain, ".", "\\.")
	}
	regexPattern := strings.Join(regexPatterns, "|")

	if isUpdate {
		return bson.M{
			"$and": []bson.M{
				{"project": bson.M{"$exists": true}},
				{"domain": bson.M{"$nin": rootDomains}},
				{"company": bson.M{"$nin": rootDomains}},
				{
					"icp": bson.M{
						"$nin": rootDomains,
						"$not": bson.M{"$regex": regexPattern, "$options": "i"},
					},
				},
			},
		}
	}

	return bson.M{
		"$or": []bson.M{
			{"domain": bson.M{"$in": rootDomains}},
			{"company": bson.M{"$in": rootDomains}},
			{"icp": bson.M{"$regex": regexPattern, "$options": "i"}},
		},
	}
}

// buildAppQuery 构建app查询条件
func (r *repository) buildAppQuery(rootDomains []string, isUpdate bool) bson.M {
	regexPatterns := make([]string, len(rootDomains))
	for i, domain := range rootDomains {
		regexPatterns[i] = "^" + strings.ReplaceAll(domain, ".", "\\.")
	}
	regexPattern := strings.Join(regexPatterns, "|")

	if isUpdate {
		return bson.M{
			"$and": []bson.M{
				{"project": bson.M{"$exists": true}},
				{"company": bson.M{"$nin": rootDomains}},
				{
					"icp": bson.M{
						"$nin": rootDomains,
						"$not": bson.M{"$regex": regexPattern, "$options": "i"},
					},
				},
			},
		}
	}

	return bson.M{
		"$or": []bson.M{
			{"name": bson.M{"$in": rootDomains}},
			{"bundleID": bson.M{"$in": rootDomains}},
			{"company": bson.M{"$in": rootDomains}},
			{"icp": bson.M{"$regex": regexPattern, "$options": "i"}},
		},
	}
}

// buildMpQuery 构建mp查询条件
func (r *repository) buildMpQuery(rootDomains []string, isUpdate bool) bson.M {
	regexPatterns := make([]string, len(rootDomains))
	for i, domain := range rootDomains {
		regexPatterns[i] = "^" + strings.ReplaceAll(domain, ".", "\\.")
	}
	regexPattern := strings.Join(regexPatterns, "|")

	if isUpdate {
		return bson.M{
			"$and": []bson.M{
				{"project": bson.M{"$exists": true}},
				{"name": bson.M{"$nin": rootDomains}},
				{"company": bson.M{"$nin": rootDomains}},
				{"bundleID": bson.M{"$nin": rootDomains}},
				{
					"icp": bson.M{
						"$nin": rootDomains,
						"$not": bson.M{"$regex": regexPattern, "$options": "i"},
					},
				},
			},
		}
	}

	return bson.M{
		"$or": []bson.M{
			{"company": bson.M{"$in": rootDomains}},
			{"icp": bson.M{"$regex": regexPattern, "$options": "i"}},
		},
	}
}
